src/back/msl.rs

```diff
@@ -403,29 +403,52 @@ impl<W: Write> Writer<W> {
                 }
             }
             crate::Expression::Constant(handle) => {
-                let kind = match module.constants[handle].inner {
+                let width = 32; //TODO: not sure how to get that...
+                match module.constants[handle].inner {
                     crate::ConstantInner::Sint(value) => {
                         write!(self.out, "{}", value)?;
-                        crate::ScalarKind::Sint
+                        Ok(MaybeOwned::Owned(
+                            crate::TypeInner::Scalar {
+                                kind: crate::ScalarKind::Sint,
+                                width,
+                            }
+                        ))
                     }
                     crate::ConstantInner::Uint(value) => {
                         write!(self.out, "{}", value)?;
-                        crate::ScalarKind::Uint
+                        Ok(MaybeOwned::Owned(
+                            crate::TypeInner::Scalar {
+                                kind: crate::ScalarKind::Uint,
+                                width,
+                            }
+                        ))
                     }
                     crate::ConstantInner::Float(value) => {
                         write!(self.out, "{}", value)?;
                         if value.fract() == 0.0 {
                             self.out.write_str(".0")?;
                         }
-                        crate::ScalarKind::Float
+                        Ok(MaybeOwned::Owned(
+                            crate::TypeInner::Scalar {
+                                kind: crate::ScalarKind::Float,
+                                width,
+                            }
+                        ))
                     }
                     crate::ConstantInner::Bool(value) => {
                         write!(self.out, "{}", value)?;
-                        crate::ScalarKind::Bool
+                        Ok(MaybeOwned::Owned(
+                            crate::TypeInner::Scalar {
+                                kind: crate::ScalarKind::Bool,
+                                width,
+                            }
+                        ))
                     }
-                };
-                let width = 32; //TODO: not sure how to get that...
-                Ok(MaybeOwned::Owned(crate::TypeInner::Scalar { kind, width }))
+                    crate::ConstantInner::Composite(handle, _) => {
+                        let inner = &module.types[handle].inner;
```

Comment:
we need to actually write down the construction of a type, i.e. something like that:
```rust
crate::ConstantInner::Composite(ref members) => {
  let ty_name = module.types[const.ty].name.or_index(const.ty);
  write!(self.out, "{}("), ty_name)?;
  for (i, &member) in members.iter().enumerate() {
    if i != 0 {
      write!(self.out, ",")?;
    }
    self.put_constant(member); // there isn't a method like this, we'd need to move the relevant part of `put_expression` into it.
  }
  wrote!(self.out, ")")?;
}
```